### PR TITLE
Discourse integration

### PR DESCRIPTION
**Pull request by cebe**
This implements SSO for Discourse as discussed in #84 

Do not merge before Discourse is not set up and deployed completely.

### CODE REVIEW

**Comment by cebe** [comment]
note to myself:

- https://meta.discourse.org/t/get-external-id-used-in-sso-through-the-api/88611

File: components/forum/DiscourseAdapter.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    public function getPostDate($user, $number)
+    {
+        if (!$user->forum_id) {
+            return false;
+        }
+
+        $tablePrefix = $this->tablePrefix;
+        $n = ((int) $number) - 1;
+        $sql = "SELECT post_date FROM {$tablePrefix}posts WHERE author_id = :user_id ORDER BY post_date ASC LIMIT " . ($n > 0 ? "$n," : '') . '1';
```
Comment (samdark) [inline components/forum/DiscourseAdapter.php]: Is there API for the purpose?

Comment (cebe) [reply components/forum/DiscourseAdapter.php]: this is still copy paste from IPB :)

getting this from the API is not that easy though.

**Comment by cebe** [comment]
TODO:

- [ ] rebase on master
- [ ] add avatar functionality to SSO

File: controllers/SiteController.php [hunk 2]
```diff
@@ -93,6 +95,76 @@ public function actionRedirect($url)
...
+            }
+        }
+
+        if (empty($url) || $url === 'index.php' || $url === 'index.php/') {
+            return $this->redirect($forumUrl);
+        }
+
+        // prevent injection of external URLs
+        if (!Url::isRelative($url)) {
+            throw new NotFoundHttpException('The requested page was not found.');
```
Comment (mdomba) [inline controllers/SiteController.php]: how about to just redirect to homepage?

Comment (cebe) [reply controllers/SiteController.php]: this is a security check to not redirect to external Urls if url is manipulated, this should never happen with real URLs.

File: components/forum/DiscourseAdapter.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+     */
+    public $apiUrl;
+    /**
+     * @var string discourse API auth token.
+     */
+    public $apiToken;
+    /**
+     * @var string discourse API user name for requests that need admin permission.
+     */
+    public $apiAdminUser = 'cebe';
```
Comment (samdark) [inline components/forum/DiscourseAdapter.php]: Hardcode?

File: components/forum/DiscourseAdapter.php [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            if (isset($userData['user']['id'])) {
+                $user->updateAttributes(['forum_id' => $userData['user']['id']]);
+            } else {
+                return 0;
+            }
+        }
+
+        $response = $this->getClient()->get([sprintf('/admin/users/%d.json', $user->forum_id), 'api_key' => $this->apiToken, 'api_username' => $this->apiAdminUser])->send();
+        if ($response->statusCode != 200) {
+            return 0;
```
Comment (samdark) [inline components/forum/DiscourseAdapter.php]: Should log failures.

File: components/forum/DiscourseAdapter.php (hunk 3 above)
Comment (cebe) [reply components/forum/DiscourseAdapter.php]: that's the default, you can override it in your config :)

File: components/forum/DiscourseAdapter.php [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    public function getPostCountsByUsername()
+    {
+        $postCounts = [];
+        $url = '/directory_items.json?period=all&order=post_count&api_key=' . urlencode($this->apiToken) . '&api_username=' . urlencode($this->apiAdminUser);
+        while (true) {
+            $response = $this->getClient()->get($url)->send();
+            if ($response->statusCode != 200) {
+                print_r($response);
```
Comment (samdark) [inline components/forum/DiscourseAdapter.php]: Should log failure instead.

File: components/forum/DiscourseAdapter.php [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+                $postCounts[$item['user']['username']] = $item['topic_count'] + $item['post_count'];
+            }
+
+            if (!isset($userData['load_more_directory_items'])) {
+                break;
+            }
+            // workaround Discourse bug https://meta.discourse.org/t/directory-items-json-api-returns-wrong-link-for-next-page/96268
+            $moreItemsUrl = str_replace('?', '.json?', $userData['load_more_directory_items']);
+            $url = $moreItemsUrl . '&api_key=' . urlencode($this->apiToken) . '&api_username=' . urlencode($this->apiAdminUser);
+            //echo count($postCounts) . '/' . $userData['total_rows_directory_items'] . ' ' . number_format(count($postCounts) / $userData['total_rows_directory_items'], 2) . "%\n";
```
Comment (samdark) [inline components/forum/DiscourseAdapter.php]: Remove?

File: components/forum/DiscourseAdapter.php [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+     * List of badges provided by the forum
+     * @return array
+     */
+    public function getForumBadges()
+    {
+        $badges = Yii::$app->cache->get('discourse_badges');
+        if ($badges === false) {
+            $response = $this->getClient()->get(['/admin/badges.json', 'api_key' => $this->apiToken, 'api_username' => $this->apiAdminUser])->send();
+            if ($response->statusCode != 200) {
+                return [];
```
Comment (samdark) [inline components/forum/DiscourseAdapter.php]: Should log failure.

File: controllers/UserController.php [hunk 8]
```diff
@@ -175,17 +176,22 @@ public function actionHalloffame()
 
     public function actionBadges()
     {
-        $badges = Badge::find()->orderBy('achieved DESC, urlname')->all();
+        $badges = Badge::find()->where(['active' => 1])->orderBy('achieved DESC, urlname')->all();
```
Comment (samdark) [inline controllers/UserController.php]: Could be moved to `active()` scope.

File: controllers/UserController.php [hunk 9]
```diff
@@ -175,17 +176,22 @@ public function actionHalloffame()
...
+            'forumBadges' => $forumBadges,
             'counts' => UserBadge::countUsers(),
         ]);
     }
 
     public function actionViewBadge($name)
     {
         /** @var Badge $badge */
-        $badge = Badge::find()->where(['urlname' => $name])->one();
+        $badge = Badge::find()->andWhere(['urlname' => $name, 'active' => 1])->one();
```
Comment (samdark) [inline controllers/UserController.php]: Could be moved to `active()` scope.

File: views/user/badges.php [hunk 10]
```diff
@@ -42,5 +45,24 @@
...
+        <div class="userbadge userbadge-forum">
+            <div class="userbadge-progress-bar" style="width: <?= round(0) ?>%"></div>
+            <div class="userbadge-info">
+                <h3>
+                    <?= Html::a(Html::encode($badge['name']), $badge['url']) ?>
+                    <?php if(isset($badge['grant_count'])): ?>
+                        <span class="x">x</span><span class="count"><?= $badge['grant_count'] ?></span>
+                    <?php endif ?>
+                </h3>
+                <p><?= $badge['description'] ?></p>
```
Comment (samdark) [inline views/user/badges.php]: Should be escaped against XSS.

Comment (cebe) [reply views/user/badges.php]: this is on purpose, description may contain links. They are managed by admins so users can not inject XSS here.

Comment (samdark) [reply views/user/badges.php]: OK.

**Pull request merged**